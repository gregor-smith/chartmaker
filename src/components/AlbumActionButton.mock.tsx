const mock: typeof import('@/components/AlbumActionButton') = {
    AlbumActionButton: props =>
        <div {...props} className='mock-album-action-button'/>
}


export const AlbumActionButton = mock.AlbumActionButton
