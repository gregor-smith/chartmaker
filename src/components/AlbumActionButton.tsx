import { h, FunctionComponent } from 'preact'
import { css } from 'emotion'


type Props = {
    onClick: () => void
    title: string
}


const style = css({
    background: 'none',
    border: 'none'
})


export const AlbumActionButton: FunctionComponent<Props> = props =>
    <button {...props} class={style}/>
