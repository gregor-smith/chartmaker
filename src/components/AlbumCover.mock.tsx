import { cx } from 'emotion'


const mock: typeof import('@/components/AlbumCover') = {
    AlbumCover: ({
        album,
        size,
        overlayClass,
        children,
        highlighted,
        ...props
    }) => {
        const json = JSON.stringify(album)
        const className = cx('mock-album-cover', overlayClass)
        return (
            <div {...props} className={className}>
                {`Album: ${json}`}
                {`Size: ${size}`}
                {`Highlighted: ${highlighted}`}
                {children}
            </div>
        )
    }
}


export const AlbumCover = mock.AlbumCover
