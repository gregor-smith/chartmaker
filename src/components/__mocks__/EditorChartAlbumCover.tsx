import type { FC } from 'react'

import type { EditorChartAlbumCoverProps } from '@/components/EditorChartAlbumCover'


export const EditorChartAlbumCover: FC<EditorChartAlbumCoverProps> = ({ album, size, highlighted }) => {
    const json = JSON.stringify(album)
    return (
        <div className='mock-editor-chart-album-cover'>
            {`Album: ${json}`}
            {`Size: ${size}`}
            {`Highlighted: ${highlighted}`}
        </div>
    )
}
