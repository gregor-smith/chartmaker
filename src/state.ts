import { CHART_ALBUMS_COUNT, DEFAULT_CHART_NAME } from './constants'


export type Album =
    | {
        placeholder: true
        id: number
    }
    | {
        placeholder: false
        id: number
        name: string
        url: string
    }


export type Chart = {
    name: string
    albums: Album[]
}


export type SearchState =
    | { tag: 'Waiting', query: string }
    | { tag: 'Loading', query: string, controller: AbortController }
    | { tag: 'Complete', query: string, albums: Album[] }
    | { tag: 'Error', query: string, message: string }


export type State = {
    apiKey: string
    charts: Chart[]
    activeChartIndex: number
    search: SearchState
    albumIDCounter: number
}


type CreateChartArguments = {
    albumIDCounter?: number
    name?: string
}


export function createChart({
    albumIDCounter = 0,
    name = DEFAULT_CHART_NAME
}: CreateChartArguments = {}): [ number, Chart ] {
    const albums: Album[] = []
    for (let id = albumIDCounter + 1; id < albumIDCounter + CHART_ALBUMS_COUNT + 1; id++) {
        albums.push({ placeholder: true, id })
    }
    return [
        albumIDCounter + CHART_ALBUMS_COUNT,
        {
            name,
            albums
        }
    ]
}


export function createInitialState(): State {
    const [ albumIDCounter, chart ] = createChart()
    return {
        apiKey: '',
        charts: [ chart ],
        activeChartIndex: 0,
        search: { tag: 'Waiting', query: '' },
        albumIDCounter
    }
}


export function escapeState(state: State): State {
    return {
        ...state,
        search: state.search.tag === 'Complete' || state.search.tag === 'Waiting'
            ? state.search
            : { tag: 'Waiting', query: state.search.query }
    }
}
