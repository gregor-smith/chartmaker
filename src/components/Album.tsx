import React, { FC } from 'react'
import { css, cx } from 'emotion'
import { Album as AlbumDetails, formatAlbumTitle } from '../state'


type Props = {
    details: AlbumDetails | null
    sizeRem: number
}


const baseStyle = css({
    backgroundColor: 'white',
    margin: '0.15rem'
})


export const Album: FC<Props> = ({ details, sizeRem }) => {
    const style = cx(
        baseStyle,
        css({
            width: `${sizeRem}rem`,
            height: `${sizeRem}rem`
        })
    )
    const image = details === null
        ? null
        : <img src={details.image.smallURL} alt={formatAlbumTitle(details)}/>
    return (
        <div className={style}>
            {image}
        </div>
    )
}
