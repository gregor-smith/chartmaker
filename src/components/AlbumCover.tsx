import type { FC, DragEvent } from 'react'
import { css, cx } from 'emotion'

import { TEXT_COLOUR, ALBUM_PADDING_SIZE, highlightBackgroundStyle } from '@/style'
import type { UnidentifiedAlbum } from '@/types'
import { unidentifiedAlbumIsPlaceholder } from '@/state'


export type AlbumCoverProps = {
    album: UnidentifiedAlbum
    size: string
    onDragStart?: (event: DragEvent<HTMLDivElement>) => void
    onDragOver?: (event: DragEvent<HTMLDivElement>) => void
    onDragEnter?: (event: DragEvent<HTMLDivElement>) => void
    onDrop?: (event: DragEvent<HTMLDivElement>) => void
    onMouseEnter?: () => void
    overlayClass?: string
    highlighted?: boolean
}


const baseContainerStyle = css({
    position: 'relative',
    background: TEXT_COLOUR,
    margin: ALBUM_PADDING_SIZE,
    cursor: 'grab'
})

const imageStyle = css({
    position: 'absolute',
    top: 0,
    left: 0,
    width: '100%',
    maxWidth: '100%',
    height: '100%',
    objectFit: 'contain'
})

const baseOverlayStyle = cx(
    imageStyle,
    css({ zIndex: 10 })
)


export const AlbumCover: FC<AlbumCoverProps> = ({
    album,
    size,
    onDragStart,
    onDragEnter,
    onDragOver,
    onDrop,
    onMouseEnter,
    children,
    overlayClass,
    highlighted
}) => {
    const containerStyle = cx(
        baseContainerStyle,
        css({
            width: size,
            height: size,
            ':not(:hover) > div': {
                display: 'none'
            }
        }),
        highlighted === false
            ? highlightBackgroundStyle
            : undefined
    )

    let image: JSX.Element | undefined
    if (!unidentifiedAlbumIsPlaceholder(album)) {
        image = (
            // lazy loading causes problems with html2canvas so keep it default
            <img className={imageStyle}
                src={album.url}
                alt={album.name}
                crossOrigin='anonymous'/>
        )
        children = (
            <div className={cx(baseOverlayStyle, overlayClass)}>
                {children}
            </div>
        )
    }

    return (
        <div className={containerStyle}
                draggable={!unidentifiedAlbumIsPlaceholder(album)}
                onDragStart={onDragStart}
                onDragEnter={onDragEnter}
                onDragOver={onDragOver}
                onDrop={onDrop}
                onMouseEnter={onMouseEnter}
                title={album?.name}>
            {image}
            {children}
        </div>
    )
}
