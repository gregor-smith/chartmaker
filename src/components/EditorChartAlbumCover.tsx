import type { FC, DragEvent } from 'react'

import type { Album } from '@/types'
import type { DispatchProps } from '@/reducer'
import { RenameAlbumButton } from '@/components/RenameAlbumButton'
import { DeleteAlbumButton } from '@/components/DeleteAlbumButton'
import { getAlbumID, identifiedAlbumIsPlaceholder } from '@/utils'
import { ChartAlbumCover } from '@/components/ChartAlbumCover'


const chartPattern = /^chart-([0-9]+)$/
const searchPattern = /^search-([0-9]+)$/


function matchDragEventData(
    event: DragEvent<HTMLDivElement>,
    pattern: RegExp
): number | null {
    for (const type of event.dataTransfer.types) {
        const match = type.match(pattern)?.[1]
        if (match !== undefined) {
            return Number(match)
        }
    }
    return null
}


function dragOver(event: DragEvent<HTMLDivElement>) {
    for (const type of event.dataTransfer.types) {
        if (type === 'Files' || searchPattern.test(type)) {
            event.dataTransfer.dropEffect = 'copy'
            event.preventDefault()
            return
        }
        if (chartPattern.test(type)) {
            event.dataTransfer.dropEffect = 'move'
            event.preventDefault()
            return
        }
    }
}


export type EditorChartAlbumCoverProps = DispatchProps & {
    album: Album
    size: string
    highlighted: number | null
}


export const EditorChartAlbumCover: FC<EditorChartAlbumCoverProps> = ({
    dispatch,
    album,
    size,
    highlighted
}) => {
    const id = getAlbumID(album)

    function dragEnter(event: DragEvent<HTMLDivElement>) {
        const sourceID = matchDragEventData(event, chartPattern)
        if (sourceID === null) {
            return
        }
        dispatch({
            tag: 'DragChartAlbum',
            sourceID,
            targetID: id
        })
        event.preventDefault()
    }

    function drop(event: DragEvent<HTMLDivElement>) {
        const file: File | undefined = event.dataTransfer.files[0]
        if (file !== undefined) {
            dispatch({
                tag: 'DropExternalFile',
                file,
                targetID: id
            })
            event.preventDefault()
            return
        }

        const sourceIndex = matchDragEventData(event, searchPattern)
        if (sourceIndex === null) {
            return
        }
        dispatch({
            tag: 'DropSearchAlbum',
            sourceIndex,
            targetID: id
        })
        event.preventDefault()
    }

    let dragStart: ((event: DragEvent<HTMLDivElement>) => void) | undefined
    let buttons: JSX.Element | undefined
    if (!identifiedAlbumIsPlaceholder(album)) {
        dragStart = event => {
            event.dataTransfer.setData(`chart-${id}`, '')
            event.dataTransfer.effectAllowed = 'copyMove'
        }
        buttons = (
            <>
                <RenameAlbumButton dispatch={dispatch} id={id}/>
                <DeleteAlbumButton dispatch={dispatch} id={id}/>
            </>
        )
    }

    return (
        <ChartAlbumCover dispatch={dispatch}
                album={album}
                size={size}
                onDragStart={dragStart}
                onDragOver={dragOver}
                onDragEnter={dragEnter}
                onDrop={drop}
                highlighted={highlighted}>
            {buttons}
        </ChartAlbumCover>
    )
}
