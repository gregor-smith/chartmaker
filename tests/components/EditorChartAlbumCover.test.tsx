import { render } from 'react-dom'
import { act } from 'react-dom/test-utils'

import type { Album } from '@/types'
import { EditorChartAlbumCover } from '@/components/EditorChartAlbumCover'
import type { Action } from '@/reducer'
import { getAlbumID } from '@/state'

import {
    RenderContainer,
    ignore,
    fireEvent,
    DragEventDataTransferMock
} from '../utils'


jest.mock('@/components/ChartAlbumCover')
jest.mock('@/components/RenameAlbumButton')
jest.mock('@/components/DeleteAlbumButton')


const container = new RenderContainer()

const dispatchMock = jest.fn<void, [ Action ]>()
afterEach(() => dispatchMock.mockClear())


const placeholderAlbum: Album = 123
const namedAlbum: Album = {
    id: 456,
    name: 'Test named album',
    url: 'https://test.com'
}


test('renders album cover for placeholder album', () => {
    render(
        <EditorChartAlbumCover dispatch={ignore}
            album={placeholderAlbum}
            size='3rem'
            highlighted/>,
        container.element
    )

    expect(container.element).toMatchSnapshot()
})


test('renders album cover with overlay buttons for named album', () => {
    render(
        <EditorChartAlbumCover dispatch={ignore}
            album={namedAlbum}
            size='5rem'
            highlighted={false}/>,
        container.element
    )

    expect(container.element).toMatchSnapshot()
})


test('starting drag sets event data for named album', () => {
    render(
        <EditorChartAlbumCover dispatch={ignore}
            album={namedAlbum}
            size='3rem'
            highlighted={undefined}/>,
        container.element
    )

    const mock = new DragEventDataTransferMock()

    act(() =>
        fireEvent(
            'dragStart',
            container.element?.firstChild,
            { dataTransfer: mock }
        )
    )

    expect(mock.setDataMock).toHaveBeenCalledTimes(1)
    expect(mock.setDataMock).toHaveBeenCalledWith(`chart-${namedAlbum.id}`, '')
    expect(mock.effectAllowedMock).toHaveBeenCalledTimes(1)
    expect(mock.effectAllowedMock).toHaveBeenCalledWith('copyMove')
})


test('starting drag does nothing for placeholder album', () => {
    render(
        <EditorChartAlbumCover dispatch={ignore}
            album={placeholderAlbum}
            size='3rem'
            highlighted={undefined}/>,
        container.element
    )

    const mock = new DragEventDataTransferMock()

    act(() =>
        fireEvent(
            'dragStart',
            container.element?.firstChild,
            { dataTransfer: mock }
        )
    )

    expect(mock.setDataMock).not.toHaveBeenCalled()
    expect(mock.effectAllowedMock).not.toHaveBeenCalled()
})


test.each([
    namedAlbum,
    placeholderAlbum
])('chart album drag enter dispatches action', album => {
    render(
        <EditorChartAlbumCover dispatch={dispatchMock}
            album={album}
            size='3rem'
            highlighted={undefined}/>,
        container.element
    )

    const dataTransferMock = new DragEventDataTransferMock([ 'chart-1' ])
    const preventDefaultMock = jest.fn<void, []>()

    act(() => {
        fireEvent(
            'dragEnter',
            container.element?.firstChild,
            {
                dataTransfer: dataTransferMock,
                preventDefault: preventDefaultMock
            }
        )
    })

    expect(dispatchMock).toHaveBeenCalledTimes(1)
    expect(dispatchMock).toHaveBeenCalledWith<[ Action ]>({
        tag: 'DragChartAlbum',
        sourceID: 1,
        targetID: getAlbumID(album)
    })
    expect(preventDefaultMock).toHaveBeenCalledTimes(1)
})


test.each([
    [ 'search-1', namedAlbum ],
    [ 'search-1', placeholderAlbum ],
    [ 'Files', namedAlbum ],
    [ 'Files', placeholderAlbum ],
    [ 'some-other-type', namedAlbum ],
    [ 'some-other-type', placeholderAlbum ],
])('any other drag enter is ignored', (type, album) => {
    render(
        <EditorChartAlbumCover dispatch={dispatchMock}
            album={album}
            size='3rem'
            highlighted={undefined}/>,
        container.element
    )

    const dataTransferMock = new DragEventDataTransferMock([ type ])
    const preventDefaultMock = jest.fn<void, []>()

    act(() => {
        fireEvent(
            'dragEnter',
            container.element?.firstChild,
            {
                dataTransfer: dataTransferMock,
                preventDefault: preventDefaultMock
            }
        )
    })

    expect(dispatchMock).toHaveBeenCalledTimes(0)
    expect(preventDefaultMock).toHaveBeenCalledTimes(0)
})


test.each([
    namedAlbum,
    placeholderAlbum
])('dragging chart album over sets move drop effect', album => {
    render(
        <EditorChartAlbumCover dispatch={ignore}
            album={album}
            size='3rem'
            highlighted={undefined}/>,
        container.element
    )

    const dataTransferMock = new DragEventDataTransferMock([ 'chart-1' ])
    const preventDefaultMock = jest.fn<void, []>()

    act(() =>
        fireEvent(
            'dragOver',
            container.element?.firstChild,
            {
                dataTransfer: dataTransferMock,
                preventDefault: preventDefaultMock
            }
        )
    )

    expect(dataTransferMock.dropEffectMock).toHaveBeenCalledTimes(1)
    expect(dataTransferMock.dropEffectMock).toHaveBeenCalledWith('move')
    expect(preventDefaultMock).toHaveBeenCalledTimes(1)
})


test.each([
    [ 'search-1', namedAlbum ],
    [ 'search-1', placeholderAlbum ],
    [ 'Files', namedAlbum ],
    [ 'Files', placeholderAlbum ]
])('dragging search album or file over sets move copy effect', (type, album) => {
    render(
        <EditorChartAlbumCover dispatch={ignore}
            album={album}
            size='3rem'
            highlighted={undefined}/>,
        container.element
    )

    const dataTransferMock = new DragEventDataTransferMock([ type ])
    const preventDefaultMock = jest.fn<void, []>()

    act(() =>
        fireEvent(
            'dragOver',
            container.element?.firstChild,
            {
                dataTransfer: dataTransferMock,
                preventDefault: preventDefaultMock
            }
        )
    )

    expect(dataTransferMock.dropEffectMock).toHaveBeenCalledTimes(1)
    expect(dataTransferMock.dropEffectMock).toHaveBeenCalledWith('copy')
    expect(preventDefaultMock).toHaveBeenCalledTimes(1)
})


test.each([
    namedAlbum,
    placeholderAlbum
])('dragging anything else over does nothing', album => {
    render(
        <EditorChartAlbumCover dispatch={ignore}
            album={album}
            size='3rem'
            highlighted={undefined}/>,
        container.element
    )

    const dataTransferMock = new DragEventDataTransferMock([ 'some-other-type' ])
    const preventDefaultMock = jest.fn<void, []>()

    act(() =>
        fireEvent(
            'dragOver',
            container.element?.firstChild,
            {
                dataTransfer: dataTransferMock,
                preventDefault: preventDefaultMock
            }
        )
    )

    expect(dataTransferMock.dropEffectMock).toHaveBeenCalledTimes(0)
    expect(preventDefaultMock).toHaveBeenCalledTimes(0)
})


test.each([
    namedAlbum,
    placeholderAlbum
])('dropping search album dispatches action', album => {
    render(
        <EditorChartAlbumCover dispatch={dispatchMock}
            album={album}
            size='3rem'
            highlighted={undefined}/>,
        container.element
    )

    const dataTransferMock = new DragEventDataTransferMock([ 'search-1' ])
    const preventDefaultMock = jest.fn<void, []>()

    act(() => {
        fireEvent(
            'drop',
            container.element?.firstChild,
            {
                dataTransfer: dataTransferMock,
                preventDefault: preventDefaultMock
            }
        )
    })

    expect(dispatchMock).toHaveBeenCalledTimes(1)
    expect(dispatchMock).toHaveBeenCalledWith<[ Action ]>({
        tag: 'DropSearchAlbum',
        sourceIndex: 1,
        targetID: getAlbumID(album)
    })
    expect(preventDefaultMock).toHaveBeenCalledTimes(1)
})


test.each([
    namedAlbum,
    placeholderAlbum
])('dropping file dispatches action', album => {
    render(
        <EditorChartAlbumCover dispatch={dispatchMock}
            album={album}
            size='3rem'
            highlighted={undefined}/>,
        container.element
    )

    const file: File = 'Test file' as any
    const dataTransferMock = new DragEventDataTransferMock(
        [ 'Files' ],
        [ file ]
    )
    const preventDefaultMock = jest.fn<void, []>()

    act(() => {
        fireEvent(
            'drop',
            container.element?.firstChild,
            {
                dataTransfer: dataTransferMock,
                preventDefault: preventDefaultMock
            }
        )
    })

    expect(dispatchMock).toHaveBeenCalledTimes(1)
    expect(dispatchMock).toHaveBeenCalledWith<[ Action ]>({
        tag: 'DropExternalFile',
        file,
        targetID: getAlbumID(album)
    })
    expect(preventDefaultMock).toHaveBeenCalledTimes(1)
})


test.each([
    [ 'chart-1', namedAlbum ],
    [ 'chart-1', placeholderAlbum ],
    [ 'some-other-type', namedAlbum ],
    [ 'some-other-type', placeholderAlbum ],
])('dropping anything else does nothing', (type, album) => {
    render(
        <EditorChartAlbumCover dispatch={dispatchMock}
            album={album}
            size='3rem'
            highlighted={undefined}/>,
        container.element
    )

    const dataTransferMock = new DragEventDataTransferMock([ type ])
    const preventDefaultMock = jest.fn<void, []>()

    act(() => {
        fireEvent(
            'drop',
            container.element?.firstChild,
            {
                dataTransfer: dataTransferMock,
                preventDefault: preventDefaultMock
            }
        )
    })

    expect(dispatchMock).toHaveBeenCalledTimes(0)
    expect(preventDefaultMock).toHaveBeenCalledTimes(0)
})


test.each([
    namedAlbum,
    placeholderAlbum
])('dispatches highlight event on mouse enter', album => {
    render(
        <EditorChartAlbumCover dispatch={dispatchMock}
            album={album}
            size='3rem'
            highlighted={undefined}/>,
        container.element
    )

    act(() => fireEvent('mouseEnter', container.element?.firstChild))

    expect(dispatchMock).toHaveBeenCalledTimes(1)
    expect(dispatchMock).toHaveBeenCalledWith<[ Action ]>({
        tag: 'HighlightAlbum',
        targetID: getAlbumID(album)
    })
})
