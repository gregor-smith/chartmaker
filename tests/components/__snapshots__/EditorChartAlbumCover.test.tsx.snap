// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders album cover for placeholder album 1`] = `
<div
  id="test-render-container"
>
  <div
    class="mock-chart-album-cover"
  >
    Album: null
    Size: 3rem
    Highlighted: true
  </div>
</div>
`;

exports[`renders album cover with overlay buttons for named album 1`] = `
<div
  id="test-render-container"
>
  <div
    class="mock-chart-album-cover"
  >
    Album: {"id":456,"name":"Test named album","url":"https://test.com"}
    Size: 5rem
    Highlighted: false
    <div
      class="mock-rename-album-button"
    >
      ID: 456
    </div>
    <div
      class="mock-delete-album-button"
    >
      ID: 456
    </div>
  </div>
</div>
`;
